File: src/buddy_cli/services.py
"""Construction of Buddy application services."""

from __future__ import annotations

from dataclasses import dataclass

from buddy_cli.config import ConfigStore
from buddy_cli.doctor import Doctor
from buddy_cli.paths import AppPaths
from buddy_cli.provisioning import Provisioner
from buddy_cli.runtime_manager import RuntimeManager
from buddy_cli.updater import Updater


@dataclass(frozen=True)
class Services:
    paths: AppPaths
    config_store: ConfigStore
    runtime_manager: RuntimeManager
    provisioner: Provisioner
    doctor: Doctor
    updater: Updater


def build_services() -> Services:
    paths = AppPaths.discover()
    config_store = ConfigStore(paths)
    runtime_manager = RuntimeManager(paths)
    return Services(
        paths=paths,
        config_store=config_store,
        runtime_manager=runtime_manager,
        provisioner=Provisioner(paths, config_store, runtime_manager),
        doctor=Doctor(paths, config_store, runtime_manager),
        updater=Updater(paths),
    )

File: src/services.rs
//! Construction of Buddy application services.

use crate::config::ConfigStore;
use crate::doctor::Doctor;
use crate::paths::AppPaths;
use crate::provisioning::Provisioner;
use crate::runtime_manager::RuntimeManager;
use crate::updater::Updater;

pub struct Services {
    pub paths: AppPaths,
    pub config_store: ConfigStore,
    pub runtime_manager: RuntimeManager,
    pub provisioner: Provisioner,
    pub doctor: Doctor,
    pub updater: Updater,
}

pub fn build_services() -> Services {
    let paths = AppPaths::discover();
    let config_store = ConfigStore::new(paths.clone());
    let runtime_manager = RuntimeManager::new(paths.clone());
    let provisioner = Provisioner::new(
        paths.clone(),
        ConfigStore::new(paths.clone()),
        RuntimeManager::new(paths.clone()),
    );
    let doctor = Doctor::new(
        paths.clone(),
        ConfigStore::new(paths.clone()),
        RuntimeManager::new(paths.clone()),
    );
    let updater = Updater::new(paths.clone());

    Services {
        paths,
        config_store,
        runtime_manager,
        provisioner,
        doctor,
        updater,
    }
}
